src/scope.rs

```diff
@@ -54,17 +54,29 @@ enum Found {
     Indirect,
 }
 
+/// Determines whether function usage matches function definitions letter-case wise
+type IsValidCasing = bool;
+
 impl<'a> Scope<'a> {
-    // fn search(&self, name: &str) -> Option<Found> {
-    fn search(&self, name: &UniCase<&str>) -> Option<Found> {
-        if self.all.contains(name) {
-            if self.defined.contains(name) || self.directly_imported.contains(name) {
-                Some(Found::Direct)
-            } else {
-                Some(Found::Indirect)
+    fn search(&self, name: &str) -> (Option<Found>, IsValidCasing) {
+        let case_insensitive_name = UniCase::new(name);
+        let direct: Set<_> = self.defined.union(&self.directly_imported).cloned().collect();
+        match direct.get(&case_insensitive_name) {
+            None => {
+                match self.all.get(&case_insensitive_name) {
+                    None => {
+                        (None, true)
+                    }
+                    Some(val) => {
+                        let is_valid_casing = val.to_string() == name;
```

Comment:
Worked quite a while to remove this duplication and tried different patter matching techniques. Have better ideas?

Reply:
> as_str

it doesn't :(

Reply:
Played around a little bit more but it only looks worse and worse...